FindCommand edit

xiinweii98 opened the pull request:
Edited FindCommand, ContainsKeywordsPredicate, StringUtil, FindCommandParser and UserGuide
- FindCommand now search for substrings
- FindCommand can take in specified prefixes and search for keywords in corresponding fields
- FindCommand can search the whole order (i.e. all field) if no prefixes are specified
- ContainsKeywordsPredicate was renamed from NameContainsKeywordsPredicate
- ContainsKeywordsPredicate now take in a HashMap mapping prefix to keywords
More details can be found in respective commit comments
Fixes #64

xiinweii98 committed:
Edited FindCommand
- FindCommand now takes in prefixes as well (user can find specific keywords from each prefix)
- FindCommand no longer require full word searches, it now includes searches of substrings

xiinweii98 committed:
Edited FindCommand
- Changed NameContainsKeywordsPredicate to ContainsKeywordsPredicate
- Added PREFIX_DELIVERY_STATUS to CliSyntax
- FindCommand can now find from delivery status as well

xiinweii98 committed:
Edit FindCommand
- Edited existing test files and cases to follow changes made to FindCommand and related classes

xiinweii98 committed:
Edit FindCommand
- Edited existing test cases

xiinweii98 committed:
Edit FindCommand
- Added more test cases to account for added prefix searching
- Added more test cases to account for changing to sub string searching

xiinweii98 committed:
Edit FindCommand
- Edited Message Usage for FindCommand
- Edited UG to reflect changes to FindCommand

pPris commented on src/main/java/seedu/cakecollate/model/order/ContainsKeywordsPredicate.java:
```diff
@@ -0,0 +1,119 @@
+                    .collect(Collectors.toSet());
+            return String.join(" ", tagsString);
+        } else if (prefix.equals(PREFIX_DATE)) {
+            LocalDate deliveryDate = order.getDeliveryDate().value;
```
I don't really understand why a delivery date has to be returned in all formats here? 😅 Also, could any of it be abstracted away in your delivery date class?

xiinweii98 replied:
Uhm because I wanted to let the user be able to search for a date using any of the format they use when they add the order + if they input the day/month/year only. Though now that I look at it I probably can do away with day/month/year since they are already included in the 4 formats provided haha. Also, yeah probably can create a method in deliverydate to get formats.

pPris commented on src/main/java/seedu/cakecollate/model/order/ContainsKeywordsPredicate.java:
```diff
@@ -0,0 +1,119 @@
+            return order.getRequest().value;
+        } else if (prefix.equals(PREFIX_DELIVERY_STATUS)) {
+            return order.getDeliveryStatus().toString();
+        } else if (prefix.equals(new Prefix("all/"))) {
```
Would using a PREFIX_ALL variable be less bug-prone in case you need to add future tests? Not very sure myself, I think as long as we assert the user doesn't use that prefix, it should be okay?

xiinweii98 replied:
Yeap my initial concern was I didn't want to let user be able to use this prefix at all thats why I only create it in the code. Just remember I can create the prefix but not set it as a prefix argmultimap accept haha thanks

pPris commented on src/main/java/seedu/cakecollate/logic/parser/FindCommandParser.java:
```diff
@@ -1,33 +1,89 @@
+                        String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+            }
+            List<String> processed = Arrays.asList(trimmedArgs.split("\\s+"));
+            prefixesToFind.put(new Prefix("all/"), processed);
```
Might want to explain the purpose of this new prefix as a comment, or maybe in the DG. Took me a bit to follow, but seems like a good design choice to have this prefix 👍

pPris commented on src/main/java/seedu/cakecollate/model/order/ContainsKeywordsPredicate.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    @Override
+    public boolean test(Order order) {
```
Seems like this method takes care of the "And" search, I'm not too sure about which part of the code takes care of the "or" search when no prefixes are specified?

xiinweii98 replied:
"or" search is done by the "all/" prefix which concatenate all the fields in an order to search the keywords against. So that becomes like alloforder.contains(keywords)

pPris reviewed (commented):
Looks good :) left a few comments about stuff I'm not able to infer from the code.

xiinweii98 committed:
Edited FindCommand based on reviews:
- Created a PREFIX_ALL
- FindCommandParser checks keywords succeeding valid and known prefixes to make sure they are non-blank
- Added tests for ^
- Abstracted getTestString for DeliveryDate
